system/serial.h

```diff
@@ -0,0 +1,181 @@
+#ifndef _SERIAL_REG_H
+#define _SERIAL_REG_H
+/**
+ * \file
+ *
+ * Provides the TTY interface. It provides an 80x25 VT100 compatible
+ * display via Serial/UART.
+ *
+ *//*
+ * Copyright (C) 2004-2022 Sam Demeulemeester.
+ */
+
+
+#define SERIAL_DEFAULT_BAUDRATE 115200
+#define SERIAL_DEFAULT_BITS     8
+#define SERIAL_DEFAULT_PARITY   0
+
+#define SERIAL_PORT_0x3F8   0
+#define SERIAL_PORT_0x2F8   1
+#define SERIAL_PORT_0x3E8   2
+#define SERIAL_PORT_0x2E8   3
+
+static const uint16_t serial_base_ports[] = { 0x3f8, 0x2f8, 0x3e8, 0x2e8 };
+
+struct serial_port {
+	bool enable;
+	bool is_mmio;
+	unsigned int parity;
+	unsigned int bits;
+	unsigned int baudrate;
+	unsigned int reg_width;
+	unsigned int refclk;
+	uintptr_t base_addr;
+};
+
+/*
+ * Definitions for VT100 commands
+ */
+
+#define TTY_CLEAR_SCREEN    "\x1b[2J"
+
+#define TTY_DISABLE_CURSOR  "\x1b[?25l"
+
+#define TTY_NORMAL          "\x1b[0m"
+#define TTY_BOLD            "\x1b[1m"
+#define TTY_UNDERLINE       "\x1b[4m"
+#define TTY_INVERSE         "\x1b[7m"
+
+/*
+ * Definitions for the Base UART Registers
+ */
+
+#define UART_REF_CLK 1843200
+
+#define UART_RX		0	/* In:  Receive buffer (DLAB=0) */
+#define UART_TX		0	/* Out: Transmit buffer (DLAB=0) */
+#define UART_DLL	0	/* Out: Divisor Latch Low (DLAB=1) */
+#define UART_DLM	1	/* Out: Divisor Latch High (DLAB=1) */
+#define UART_IER	1	/* Out: Interrupt Enable Register */
+#define UART_IIR	2	/* In:  Interrupt ID Register */
+#define UART_FCR	2	/* Out: FIFO Control Register */
+#define UART_EFR	2	/* I/O: Extended Features Register */
+				/* (DLAB=1, 16C660 only) */
+#define UART_LCR	3	/* Out: Line Control Register */
+#define UART_MCR	4	/* Out: Modem Control Register */
+#define UART_LSR	5	/* In:  Line Status Register */
+#define UART_MSR	6	/* In:  Modem Status Register */
+#define UART_SCR	7	/* I/O: Scratch Register */
+
+/*
+ * Definitions for the Line Control Register
+ */
+
+#define UART_LCR_DLAB	0x80	/* Divisor latch access bit */
+#define UART_LCR_SBC	0x40	/* Set break control */
+#define UART_LCR_SPAR	0x20	/* Stick parity (?) */
+#define UART_LCR_EPAR	0x10	/* Even parity select */
+#define UART_LCR_PARITY	0x08	/* Parity Enable */
+#define UART_LCR_STOP	0x04	/* Stop bits: 0=1 stop bit, 1= 2 stop bits */
+#define UART_LCR_WLEN5  0x00	/* Wordlength: 5 bits */
+#define UART_LCR_WLEN6  0x01	/* Wordlength: 6 bits */
+#define UART_LCR_WLEN7  0x02	/* Wordlength: 7 bits */
+#define UART_LCR_WLEN8  0x03	/* Wordlength: 8 bits */
+
+/*
+ * Definitions for the Line Status Register
+ */
+#define UART_LSR_TEMT	0x40	/* Transmitter empty */
+#define UART_LSR_THRE	0x20	/* Transmit-hold-register empty */
+#define UART_LSR_BI     0x10	/* Break interrupt indicator */
```

Comment:
Whitespace here and in the following code blocks.

Reply:
PS: Most of these blank lines are intentional. They basically increase readability by separating code blocks (or isolate important line visually). Removing all blank lines add a lot of obfuscation to the code.

Reply:
The blank lines are fine indeed, but there's inconsistent indentation of the right side of #defines :)